### PR TITLE
feat(groups): add typed membership request variants (correct wire format)

### PR DESCRIPTION
**Pull request by Salientekill**
## Summary

Adds typed `GroupNotificationAction` variants for the three separate membership request notification tags from `WAWebHandleGroupNotificationConst.GROUP_NOTIFICATION_TAG`, replacing the previous incorrect bifurcation on a non-existent `membership_approval_request_response` child.

### New variants

- **`MembershipApprovalRequest { request_method, parent_group_jid }`** — `<membership_approval_request>` — a user requested to join; flat attrs only, requester is `GroupNotification::participant`
- **`CreatedMembershipRequests { request_method, parent_group_jid, requests }`** — `<created_membership_requests>` — admin-side: new pending requests appeared; uses `<requested_user>` children
- **`RevokedMembershipRequests { participants }`** — `<revoked_membership_requests>` — requests rejected by admin or cancelled by requester; uses `<participant jid=.../>` children

### New supporting type

**`MembershipRequestMethod`** — maps `request_method` attr to `InviteLink | LinkedGroupJoin | NonAdminAdd` (mirrors `WAWebRequestMethodType`; defaults to `InviteLink` matching WA Web's fallback)

### Notes

- Approval of a request results in a regular `<add>` notification — no new variant needed
- `membership_approval_request_response` child does not appear in captured WA Web JS (confirmed by @jlucaso1)
- `Unknown` catch-all preserved for forward compatibility; uses `node.tag.to_string()` for consistency
- 12 tests covering all three new variants plus existing actions

Co-authored-by: Salientekill

## Summary by CodeRabbit

* **New Features**
  * Group membership notifications now cover submission, batch-created, and revoked membership requests; notifications surface parent-group and participant details and include a request-method indicator with a sensible default when not provided.

* **Tests**
  * Added unit tests validating parsing and handling of membership approval requests, created membership requests, revoked membership requests, and missing optional parent-group data.

### CODE REVIEW

File: wacore/src/stanza/groups.rs [hunk 1]
```diff
@@ -14,6 +14,18 @@ use serde::Serialize;
...
 use wacore_binary::{Node, NodeRef};
 
+/// Whether a membership request was approved or rejected.
+///
+/// Used in [`GroupNotificationAction::MembershipRequestAction`].
+/// The admin who performed the action is available on the parent
+/// [`GroupNotification::participant`] field.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
+#[serde(rename_all = "lowercase")]
+pub enum MembershipRequestActionKind {
```
Comment (jlucaso1) [inline wacore/src/stanza/groups.rs]: This enum assumes an Approved/Rejected model, but WA Web doesn't send approve/reject through group notifications. Instead, the `membership_approval_request` tag carries a `request_method` attribute.

From `WAWebHandleGroupNotification`:
```js
requestMethod: E[a.attrString("request_method")] ?? RequestMethod.InviteLink
```

The `RequestMethod` enum (from `WAWebRequestMethodType`) is:
- `InviteLink`
- `LinkedGroupJoin`
- `NonAdminAdd`

Suggestion: replace this with something like:
```rust
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum MembershipRequestMethod {
    InviteLink,
    LinkedGroupJoin,
    NonAdminAdd,
}
```

File: wacore/src/stanza/groups.rs [hunk 2]
```diff
@@ -113,6 +125,20 @@ pub enum GroupNotificationAction {
     },
     /// `<membership_approval_mode><group_join state="on|off"/></membership_approval_mode>`
     MembershipApprovalMode { enabled: bool },
+    /// `<membership_approval_request>` with `<membership_approval_request_response>` child —
```
Comment (jlucaso1) [inline wacore/src/stanza/groups.rs]: The `membership_approval_request_response` child doesn't exist in captured WA Web JS (zero matches across all files). Neither whatsmeow nor Baileys parse it.

WA Web treats these as three separate notification tags with different structures. Here's what the variants should look like to match:

```rust
/// `<membership_approval_request request_method="..." parent_group_jid="..."/>`
/// User requested to join. Requester is on parent `GroupNotification::participant`.
MembershipApprovalRequest {
    request_method: MembershipRequestMethod,
    parent_group_jid: Option<Jid>,
},
/// `<created_membership_requests request_method="..." parent_group_jid="...">` 
/// Admin-side: new join requests appeared.
CreatedMembershipRequests {
    request_method: MembershipRequestMethod,
    parent_group_jid: Option<Jid>,
    /// `<requested_user>` children (note: not `<participant>`)
    requests: Vec<GroupParticipantInfo>,
},
/// `<revoked_membership_requests>` 
/// Requests rejected by admin or cancelled by requester.
RevokedMembershipRequests {
    participants: Vec<Jid>,
},
```

Note that `created_membership_requests` uses `<requested_user>` children (with `jid`, `username`, `phone_number` attrs), while `revoked_membership_requests` uses `<participant>` children (just `jid`).

File: wacore/src/stanza/groups.rs [hunk 3]
```diff
@@ -307,6 +335,31 @@ fn parse_action(node: &NodeRef<'_>) -> Option<GroupNotificationAction> {
                 .is_some_and(|s| s == "on");
             GroupNotificationAction::MembershipApprovalMode { enabled }
         }
+        "membership_approval_request" => {
```
Comment (jlucaso1) [inline wacore/src/stanza/groups.rs:342]: The bifurcation on `membership_approval_request_response` doesn't match WA Web. In `WAWebHandleGroupNotification`, this tag is handled as a simple flat node with just attributes (no children parsed at all).

Also, WA Web's model updater (`WAWebUpdateModelForGroupAction`) does nothing with this action, it's just `break;`. It only generates a system message.

Here's a sketch for the three match arms:

```rust
"membership_approval_request" => {
    let request_method = parse_request_method(node);
    let parent_group_jid = node.attrs().optional_jid("parent_group_jid");
    GroupNotificationAction::MembershipApprovalRequest {
        request_method,
        parent_group_jid,
    }
}
"created_membership_requests" => {
    let request_method = parse_request_method(node);
    let parent_group_jid = node.attrs().optional_jid("parent_group_jid");
    // WA Web uses <requested_user> children, not <participant>
    let requests = parse_requested_users(node);
    GroupNotificationAction::CreatedMembershipRequests {
        request_method,
        parent_group_jid,
        requests,
    }
}
"revoked_membership_requests" => {
    let participants = parse_participant_jids(node);
    GroupNotificationAction::RevokedMembershipRequests { participants }
}
```

You'd need a small `parse_request_method` helper that reads the `request_method` attr and maps `"invite_link"` / `"linked_group_join"` / `"non_admin_add"` to the enum, defaulting to `InviteLink` (matching WA Web's fallback).

File: wacore/src/stanza/groups.rs [hunk 4]
```diff
@@ -307,6 +335,31 @@ fn parse_action(node: &NodeRef<'_>) -> Option<GroupNotificationAction> {
...
+            if let Some(response) = node.get_optional_child("membership_approval_request_response")
+            {
+                // Admin acted on pending requests (approved or rejected).
+                let action = match response.attrs().optional_string("action").as_deref() {
+                    Some("approve") => MembershipRequestActionKind::Approved,
+                    Some("reject") => MembershipRequestActionKind::Rejected,
+                    // Unknown or missing action — preserve for forward compatibility.
+                    _ => {
+                        return Some(GroupNotificationAction::Unknown {
+                            tag: "membership_approval_request".to_string(),
```
Comment (jlucaso1) [inline wacore/src/stanza/groups.rs]: Nit: this hardcodes the string literal. For consistency with the `other` catch-all below (which does `other.to_string()`), prefer `node.tag.to_string()` here. Avoids the duplication and stays consistent if the tag ever gets renamed.

File: wacore/src/stanza/groups.rs [hunk 5]
```diff
@@ -592,6 +645,120 @@ mod tests {
         }
     }
 
+    #[test]
```
Comment (jlucaso1) [inline wacore/src/stanza/groups.rs:689]: Tests are well-structured, nice coverage of the happy paths and the unknown-action edge case. They'll just need updating once the wire format changes.

For example, the `membership_approval_request` test would look something like:
```rust
#[test]
fn test_parse_membership_approval_request() {
    let node = make_notification(vec![
        NodeBuilder::new("membership_approval_request")
            .attr("request_method", "invite_link")
            .build(),
    ]);
    let notif = GroupNotification::try_from_node_ref(&node.as_node_ref()).unwrap();
    match &notif.actions[0] {
        GroupNotificationAction::MembershipApprovalRequest {
            request_method, parent_group_jid
        } => {
            assert_eq!(*request_method, MembershipRequestMethod::InviteLink);
            assert!(parent_group_jid.is_none());
        }
        other => panic!("expected MembershipApprovalRequest, got {:?}", other),
    }
}
```

And then separate tests for `created_membership_requests` (with `<requested_user>` children) and `revoked_membership_requests` (with `<participant>` children).

**Review by jlucaso1 — CHANGES_REQUESTED** [review]
Hey, nice work structuring this with proper tests and the `Unknown` fallback for forward compat. The intent is solid.

However, I checked the captured JS and the wire format here doesn't match what WA Web actually does. The main issue is that WA Web doesn't use a single overloaded `membership_approval_request` tag. It uses **three separate top-level tags** (from `WAWebHandleGroupNotificationConst.GROUP_NOTIFICATION_TAG`):

1. `membership_approval_request` - a user requested to join (system message only, parsed attrs: `request_method`, `parent_group_jid`)
2. `created_membership_requests` - admin-side: new requests appeared (children: `<requested_user jid=... username=... phone_number=.../>`)
3. `revoked_membership_requests` - requests were rejected/cancelled (children: `<participant jid=.../>`)

Also, `membership_approval_request_response` as a child node doesn't appear anywhere in the captured JS, and neither whatsmeow nor Baileys parse it either. The approval of a request results in a regular `<add>` notification, not a membership request notification.

I left inline comments with suggestions for each piece. The tests are well-structured so they should be easy to adapt once the wire format is corrected.

**Review by jlucaso1 — APPROVED** [review]
LGTM thanks!
